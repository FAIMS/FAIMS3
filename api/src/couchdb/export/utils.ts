import {
  Annotations,
  FAIMSAttachmentReference,
  FieldSummary,
} from '@faims3/data-model';
import {generateFilenameForAttachment} from './attachmentExport';

/**
 * TODO it would be better to merge the header generation and value formatting
 * logic together - currently some parts are encoded twice in two different
 * ways.
 *
 *
 * generate a suitable value for the CSV export from a field value.  Serialise
 * filenames, gps coordinates, etc.
 */
export const csvFormatValue = (
  fieldName: string,
  fieldType: string,
  value: any,
  hrid: string,
  filenames: string[],
  viewsetId: string
) => {
  const result: {[key: string]: any} = {};
  if (fieldType === 'faims-attachment::Files') {
    if (value instanceof Array) {
      if (value.length === 0) {
        result[fieldName] = '';
        return result;
      }
      const valueList = (value as FAIMSAttachmentReference[])
        .filter(f => !!f.file_type)
        .map((fInfo: FAIMSAttachmentReference) => {
          const filename = generateFilenameForAttachment({
            fieldId: fieldName,
            hrid,
            viewID: viewsetId,
            // Pass in the file type
            fileMimeType: fInfo.file_type,
            filenames,
          });
          filenames.push(filename);
          return filename;
        });
      result[fieldName] = valueList.join(';');
    } else {
      result[fieldName] = value;
    }
    return result;
  }

  // gps locations
  if (fieldType === 'faims-pos::Location') {
    if (
      value instanceof Object &&
      'geometry' in value &&
      value.geometry.coordinates.length === 2
    ) {
      result[fieldName] = value;
      result[fieldName + '_latitude'] = value.geometry.coordinates[1];
      result[fieldName + '_longitude'] = value.geometry.coordinates[0];
      result[fieldName + '_accuracy'] = value.properties.accuracy || '';
    } else {
      result[fieldName] = value;
      result[fieldName + '_latitude'] = '';
      result[fieldName + '_longitude'] = '';
      result[fieldName + '_accuracy'] = '';
    }
    return result;
  }

  if (fieldType === 'faims-core::JSON') {
    // map location, if it's a point we can pull out lat/long
    if (
      value instanceof Object &&
      'features' in value &&
      value.features.length > 0 &&
      value.features[0]?.geometry?.type === 'Point' &&
      value.features[0].geometry.coordinates.length === 2
    ) {
      result[fieldName] = value;
      result[fieldName + '_latitude'] =
        value.features[0].geometry.coordinates[1];
      result[fieldName + '_longitude'] =
        value.features[0].geometry.coordinates[0];
      return result;
    } else {
      result[fieldName] = value;
      result[fieldName + '_latitude'] = '';
      result[fieldName + '_longitude'] = '';
    }
  }

  if (fieldType === 'faims-core::Relationship') {
    if (value instanceof Array) {
      result[fieldName] = value
        .map((v: any) => {
          const relation_name = v.relation_type_vocabPair
            ? v.relation_type_vocabPair[0]
            : 'unknown relation';
          return `${relation_name}/${v.record_id}`;
        })
        .join(';');
    } else {
      result[fieldName] = value;
    }
    return result;
  }

  // default to just the value
  result[fieldName] = value;
  return result;
};

/**
 * Convert annotations on a field to a format suitable for CSV export
 */
export const csvFormatAnnotation = (
  field: FieldSummary,
  {annotation, uncertainty}: Annotations
) => {
  const result: {[key: string]: any} = {};
  if (field.annotation !== '')
    result[field.name + '_' + field.annotation] = annotation;
  if (field.uncertainty !== '')
    result[field.name + '_' + field.uncertainty] = uncertainty
      ? 'true'
      : 'false';
  return result;
};

/**
 * Format the data for a single record for CSV export
 *
 * @returns a map of column headings to values
 */
export const convertDataForOutput = (
  fields: FieldSummary[],
  data: any,
  annotations: {[name: string]: Annotations},
  hrid: string,
  filenames: string[],
  viewsetId: string
) => {
  let result: {[key: string]: any} = {};
  fields.map((field: any) => {
    if (field.name in data) {
      const formattedValue = csvFormatValue(
        field.name,
        field.type,
        data[field.name],
        hrid,
        filenames,
        viewsetId
      );
      const formattedAnnotation = csvFormatAnnotation(
        field,
        annotations[field.name] || {}
      );
      result = {...result, ...formattedValue, ...formattedAnnotation};
    } else {
      console.error('field missing in data', field.name, data);
    }
  });
  return result;
};
