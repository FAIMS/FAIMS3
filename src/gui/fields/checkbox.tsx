/*
 * Copyright 2021, 2022 Macquarie University
 *
 * Licensed under the Apache License Version 2.0 (the, "License");
 * you may not use, this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
 * See, the License, for the specific language governing permissions and
 * limitations under the License.
 *
 * Filename: checkbox.tsx
 * Description:
 *   TODO
 */

import React from 'react';
import MuiCheckbox from '@mui/material/Checkbox';
import FormControl from '@mui/material/FormControl';
import {
  FormControlLabel,
  FormHelperText,
  FormHelperTextProps,
  FormControlLabelProps,
} from '@mui/material';
import {fieldToCheckbox, CheckboxProps} from 'formik-mui';

interface Props {
  FormControlLabelProps?: FormControlLabelProps; // deprecated
  FormHelperTextProps?: FormHelperTextProps; // deprecated
  helperText?: string;
  label?: string;
}

export class Checkbox extends React.Component<CheckboxProps & Props> {
  render() {
    const {
      FormControlLabelProps,
      FormHelperTextProps,
      helperText,
      ...checkboxWithLabelProps
    } = this.props;

    // for backwards compatibility we check these properties as well as
    // just the plain label and helperText properties
    const label =
      this.props.label || FormControlLabelProps?.label || this.props.field.name;

    const theHelperText = helperText || FormHelperTextProps?.children || '';

    let error = false;
    if (
      checkboxWithLabelProps.form.errors[checkboxWithLabelProps.field.name] &&
      checkboxWithLabelProps.form.touched[checkboxWithLabelProps.field.name]
    ) {
      error = true;
    }

    return (
      <FormControl error={error}>
        <FormControlLabel
          label={label}
          control={
            <MuiCheckbox
              {...fieldToCheckbox(checkboxWithLabelProps)}
              checked={checkboxWithLabelProps.field.value}
            />
          }
        />
        {error ? (
          <FormHelperText
            children={
              checkboxWithLabelProps.form.errors[
                checkboxWithLabelProps.field.name
              ] as string
            }
          />
        ) : (
          <FormHelperText children={theHelperText} />
        )}
      </FormControl>
    );
  }
}

// const uiSpec = {
//   'component-namespace': 'faims-custom', // this says what web component to use to render/acquire value from
//   'component-name': 'Checkbox',
//   'type-returned': 'faims-core::Bool', // matches a type in the Project Model
//   'component-parameters': {
//     name: 'checkbox-field',
//     id: 'checkbox-field',
//     required: false,
//     type: 'checkbox',
//     label: 'Terms and Conditions',
//     helperText: 'Read the terms and conditions carefully.',

//     // Label: {label: 'Terms and Conditions'},
//   },
//   validationSchema: [
//     ['yup.bool'],
//     // ['yup.oneOf', [true], ''],
//   ],
//   initialValue: false,
// };
