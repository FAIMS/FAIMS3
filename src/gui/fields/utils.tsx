/*
 * Copyright 2021, 2022 Macquarie University
 *
 * Licensed under the Apache License Version 2.0 (the, "License");
 * you may not use, this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
 * See, the License, for the specific language governing permissions and
 * limitations under the License.
 *
 * Filename: utils.tsx
 * Description:
 *   TODO
 */

import React from 'react';
import {Values, initialValues} from './RichText.test';
import {getComponentFromFieldConfig} from '../components/record/fields';
import {render} from '@testing-library/react';
import {Formik, FormikConfig, FormikProps} from 'formik';

const customRender = (
  ui: React.ReactElement,
  formikOpts: Omit<Props, 'children'>,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, {
    wrapper: function CustomRendererWrapper(props) {
      return <FormikWrapper {...formikOpts} {...props} />;
    },
    ...options,
  });
};

/**
 * Render a form element via Formik for testing
 * @param ui - the form element to render
 * @param props - properties to inject into the element
 **/
export const renderForm = (
  ui?: React.ReactNode,
  props?: Partial<FormikConfig<Values>>
) => {
  let injected: FormikProps<Values>;
  const {rerender, ...rest} = render(
    <Formik onSubmit={() => {}} initialValues={initialValues} {...props}>
      {(formikProps: FormikProps<Values>) =>
        (injected = formikProps) && ui ? ui : null
      }
    </Formik>
  );

  return {
    getFormProps(): FormikProps<Values> {
      return injected;
    },
    ...rest,
    rerender: () =>
      rerender(
        <Formik onSubmit={() => {}} initialValues={initialValues} {...props}>
          {(formikProps: FormikProps<Values>) =>
            (injected = formikProps) && ui ? ui : null
          }
        </Formik>
      ),
  };
};
/**
 * instantiateField - instantiate a field from a uiSpec for testing
 * @param uiSpec - uiSpec for the field we want to render
 * @returns - the rendered field element
 */
export const instantiateField = (uiSpec: any) => {
  const formProps = {
    values: {},
    errors: {},
    touched: {},
    handleChange: () => {},
    setFieldValue: () => {},
  };
  const element = getComponentFromFieldConfig(uiSpec, 'test', formProps);
  return renderForm(element);
};

export * from '@testing-library/react';

export {customRender as render};
