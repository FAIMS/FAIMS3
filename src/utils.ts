/*
 * Copyright 2021, 2022 Macquarie University
 *
 * Licensed under the Apache License Version 2.0 (the, "License");
 * you may not use, this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
 * See, the License, for the specific language governing permissions and
 * limitations under the License.
 *
 * Filename: utils.ts
 * Description:
 *   Contains utility functions which lack a better location.
 */
import {Device} from '@capacitor/device';
import {Filesystem, Directory} from '@capacitor/filesystem';

/// Downloads a blob as a file onto a user's device
export function downloadBlob(b: Blob, filename: string) {
  const u = URL.createObjectURL(b);
  const a = document.createElement('a');
  a.href = u;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(u);
}

export async function downloadStringOnApp(
  s: string,
  encoding: string,
  filename: string
) {
  const info = await Device.getInfo();
  if (info.platform === 'ios') {
    await Filesystem.writeFile({
      path: filename,
      data: s,
      directory: Directory.Documents,
      encoding: encoding,
    });
  } else if (info.platform === 'android') {
    const cache_res = await Filesystem.writeFile({
      path: filename,
      data: s,
      directory: Directory.Cache,
      encoding: encoding,
    });
    await Mediastore.saveToDownloads({
      path: filename,
      directory: cache_res.url,
    });
    await Filesystem.deleteFile({path: filename, directory: Directory.Cache});
  } else {
    throw Error('Do not know how to save on this platform');
  }
}
