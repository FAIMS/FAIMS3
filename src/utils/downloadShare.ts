/*
 * Copyright 2021, 2022 Macquarie University
 *
 * Licensed under the Apache License Version 2.0 (the, "License");
 * you may not use, this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
 * See, the License, for the specific language governing permissions and
 * limitations under the License.
 *
 * Filename: utils.ts
 * Description:
 *   Contains utility functions related to downloading and sharing
 */
import {Filesystem, Directory, Encoding} from '@capacitor/filesystem';
import {Share} from '@capacitor/share';

/// Downloads a blob as a file onto a user's device
export function downloadBlob(b: Blob, filename: string) {
  const u = URL.createObjectURL(b);
  const a = document.createElement('a');
  a.href = u;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(u);
}

export async function shareStringAsFileOnApp(
  s: string,
  title: string,
  dialogTitle: string,
  filename: string
) {
  const isodate = new Date().toJSON().slice(0, 10);
  console.debug('Starting writing of file');
  for (const dir of [
    Directory.Library,
    Directory.Cache,
    Directory.Documents,
    Directory.Data,
    Directory.External,
    Directory.ExternalStorage,
  ]) {
    try {
      console.debug('Trying ', dir);
      const url = (
        await Filesystem.writeFile({
          path: `${isodate}-${dir}-${filename}`,
          data: s,
          directory: dir,
          encoding: Encoding.UTF8,
          recursive: true,
        })
      ).uri;
      console.debug('Writing of file complete, sharing file', url);
      await Share.share({
        title: `${title} ${dir} ${isodate}.json`,
        text: dialogTitle,
        url: url,
        dialogTitle: dialogTitle,
      });
      break;
    } catch (err) {
      console.error('Sharing failed with', dir, err);
    }
  }
  console.debug('Shared file');
}
