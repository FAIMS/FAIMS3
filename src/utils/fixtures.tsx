import {v4 as uuidv4} from 'uuid';

export function related_links_from_fields(
  record_id: string,
  hrid: string,
  record_type: string
) {
  return [
    {
      record_id: record_id,
      hrid: hrid,
      type: record_type,
      route: 'route',
      relation_type_vocabPair: ['is child of', 'has parent'],
      link: {
        record_id: uuidv4(),
        hrid: 'Artefact-V3.4',
        type: 'Category-X',
        route: 'linked record route',
        section: 'Section Pass 1',
        field_id: uuidv4(),
        field_label: 'Container',
      },
      lastUpdatedBy: '10/12/2020 11:09am by John Smith',
    },
    {
      record_id: record_id,
      hrid: hrid,
      type: record_type,
      route: 'route',
      relation_type_vocabPair: ['is related to', 'is related to'],
      link: {
        record_id: uuidv4(),
        hrid: 'Mundi-V3.4',
        type: 'Water',
        route: 'linked record route',
        section: 'Section 1',
        field_id: uuidv4(),
        field_label: 'Sample ID',
      },
      lastUpdatedBy: '10/12/2020 11:09am by Chris Smith',
    },
    {
      record_id: uuidv4(),
      hrid: 'Feature-001',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['is child of', 'has parent'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section A',
        field_id: uuidv4(),
        field_label: 'EH',
      },
      lastUpdatedBy: '01/09/2022 10:34am by Jo Jones',
    },
    {
      record_id: uuidv4(),
      hrid: 'Feature-002',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['is child of', 'has parent'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section B',
        field_id: uuidv4(),
        field_label: 'PH',
      },
      lastUpdatedBy: '01/08/2022 8:34am by Jo Jones',
    },
    {
      record_id: uuidv4(),
      hrid: 'Feature-001',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['contains', 'is contained by'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section B',
        field_id: uuidv4(),
        field_label: 'PH',
      },
      lastUpdatedBy: '20/08/2022 9:34am by Jo Jones',
    },
  ];
}
export function field_level_links(
  record_id: string,
  hrid: string,
  record_type: string
) {
  return [
    {
      record_id: uuidv4(),
      hrid: 'Feature-001',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['is child of', 'has parent'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section A',
        field_id: uuidv4(),
        field_label: 'EH',
      },
      lastUpdatedBy: '01/09/2022 10:34am by Jo Jones',
    },
    {
      record_id: uuidv4(),
      hrid: 'Feature-002',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['is child of', 'has parent'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section A',
        field_id: uuidv4(),
        field_label: 'EH',
      },
      lastUpdatedBy: '01/08/2022 8:34am by Jo Jones',
    },
    {
      record_id: uuidv4(),
      hrid: 'Feature-001',
      type: 'Artefact',
      route: 'linked record route',
      relation_type_vocabPair: ['is contained by', 'contains'],
      link: {
        record_id: record_id,
        hrid: hrid,
        type: record_type,
        route: 'route',
        section: 'Section A',
        field_id: uuidv4(),
        field_label: 'EH',
      },
      lastUpdatedBy: '20/08/2022 9:34am by Jo Jones',
    },
  ];
}
