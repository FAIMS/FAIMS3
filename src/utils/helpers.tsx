import {TokenContents} from '../datamodel/core';

export function tokenExists(token: null | undefined | TokenContents) {
  return token !== null && token !== undefined;
}

export function tokenValid(token: null | undefined | TokenContents) {
  /**
   * Check for expiry AND validity
   */
  // TODO
  return true;
}

export function checkToken(token: null | undefined | TokenContents) {
  /**
   * Check if the token exists, and whether it's valid
   */
  return tokenExists(token) && tokenValid(token);
}
