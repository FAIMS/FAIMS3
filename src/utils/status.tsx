import {ActionType} from '../context/actions';
import {ContextType} from '../context/store';
import {SyncStatusCallbacks} from 'faims3-datamodel';

export type SyncActionTypes =
  | ActionType.IS_SYNCING_DOWN
  | ActionType.IS_SYNCING_UP;

export function startSync(
  dispatch: ContextType['dispatch'],
  syncAction: SyncActionTypes
) {
  /**
   * Dispatch IS_SYNCING_UP or IS_SYNCING_DOWN =true
   * Set to false after 5 seconds
   */
  const controller = new AbortController();

  dispatch({type: syncAction, payload: true});
  const apiTimeout = setTimeout(() => {
    dispatch({type: syncAction, payload: false});
    clearTimeout(apiTimeout);
    controller.abort();
  }, 5000);

  return;
}

function sendErrorNotification(
  dispatch: ContextType['dispatch'],
  message: string
) {
  dispatch({
    type: ActionType.ADD_ALERT,
    payload: {message: message, severity: 'error'},
  });
}

export function setSyncError(
  dispatch: ContextType['dispatch'],
  has_error: boolean
) {
  /**
   * Toggle the sync error state
   */
  dispatch({type: ActionType.IS_SYNC_ERROR, payload: has_error});
}

export function getSyncStatusCallbacks(
  dispatch: ContextType['dispatch']
): SyncStatusCallbacks {
  const handleStartSyncUp = () => {
    startSync(dispatch, ActionType.IS_SYNCING_UP);
  };
  const handleStartSyncDown = () => {
    startSync(dispatch, ActionType.IS_SYNCING_DOWN);
  };
  const handleStartSyncError = () => {
    setSyncError(dispatch, true);
  };
  const handleStartSyncDenied = () => {
    // TODO: Add denied status
    sendErrorNotification(
      dispatch,
      'Sync Authentication Error: try refreshing your roles'
    );
    setSyncError(dispatch, true);
  };
  return {
    sync_up: handleStartSyncUp,
    sync_down: handleStartSyncDown,
    sync_error: handleStartSyncError,
    sync_denied: handleStartSyncDenied,
  };
}
