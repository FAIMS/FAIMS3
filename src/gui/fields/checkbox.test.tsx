import React from 'react';
import {Field} from 'formik';
import * as Yup from 'yup';
import {render, fireEvent} from './utils';

import {Checkbox} from './checkbox';

test('renders as formik component', () => {
  const onSubmit = jest.fn();
  const {asFragment} = render(
    <Field component={Checkbox} name="checked" type="checkbox" />,
    {
      initialValues: {
        checked: false,
      },
      onSubmit,
    }
  );

  expect(asFragment()).toMatchSnapshot();
});

test('submitting checked is true', async () => {
  // given
  const onSubmit = jest.fn();
  const {getByTestId, findByText} = render(
    <Field
      component={Checkbox}
      name="checked"
      label="Checkbox"
      type="checkbox"
      inputProps={{
        'data-testid': 'checkbox',
      }}
    />,
    {
      onSubmit,
      initialValues: {
        checked: true,
      },
    }
  );

  const input = getByTestId('checkbox');
  const submit = getByTestId('submit');

  // when
  fireEvent.change(input, {target: {value: 'checked'}});
  fireEvent.click(submit);

  await findByText('submitted');

  // then
  expect(onSubmit).toBeCalledWith(
    {
      checked: true,
    },
    expect.anything()
  );
});

test('pass FormControlLabelProps and FormHelperTextProps (checked false)', async () => {
  // given
  const onSubmit = jest.fn();
  const {getByTestId, findByText, asFragment} = render(
    <Field
      component={Checkbox}
      name="checked"
      label="Checkbox"
      type="checkbox"
      inputProps={{
        'data-testid': 'checkbox',
      }}
      FormControlLabelProps={{label: 'Terms and Conditions'}}
      FormHelperTextProps={{
        children: 'Read the terms and conditions carefully.',
      }}
    />,
    {
      onSubmit,
      initialValues: {
        checked: false,
      },
    }
  );

  expect(asFragment()).toMatchSnapshot();

  const input = getByTestId('checkbox');
  const submit = getByTestId('submit');

  // when
  fireEvent.change(input, {target: {value: 'checked'}});
  fireEvent.click(submit);

  await findByText('submitted');
  await findByText('Terms and Conditions');
  await findByText('Read the terms and conditions carefully.');

  // then
  expect(onSubmit).toBeCalledWith(
    {
      checked: false,
    },
    expect.anything()
  );
});

test('displays error if field is touched and checkbox is false, if validation schema requires it to be true', async () => {
  const validationSchema = Yup.object().shape({
    checked: Yup.bool()
      .oneOf([true], 'You must accept the terms and conditions')
      .required(),
  });

  const onSubmit = jest.fn();
  const {getByTestId, findByText, asFragment} = render(
    <Field
      component={Checkbox}
      name="checked"
      label="Checkbox"
      type="checkbox"
      inputProps={{
        'data-testid': 'checkbox',
      }}
      FormControlLabelProps={{label: 'Terms and Conditions'}}
      FormHelperTextProps={{
        children: 'Read the terms and conditions carefully.',
      }}
    />,
    {
      onSubmit,
      initialValues: {
        checked: false,
      },
      validationSchema: validationSchema,
    }
  );

  expect(asFragment()).toMatchSnapshot();

  const input = getByTestId('checkbox');

  // Call blur without inputting anything which should trigger a validation error
  fireEvent.blur(input);

  // The validation message should be shown and the label should still be present
  await findByText('Terms and Conditions');
  // check for Mui-error in classList
  const errorText = await findByText(
    'You must accept the terms and conditions'
  );
  expect(errorText).toHaveClass('Mui-error');
});
