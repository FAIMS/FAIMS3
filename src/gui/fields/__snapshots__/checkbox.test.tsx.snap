// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays error if field is touched and checkbox is false, if validation schema requires it to be true 1`] = `
<DocumentFragment>
  <form
    action="#"
    data-testid="form"
  >
    <div
      class="MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          label="Checkbox"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-16"
              data-indeterminate="false"
              data-testid="checkbox"
              name="checked"
              type="checkbox"
              value="false"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Terms and Conditions
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      >
        Read the terms and conditions carefully.
      </p>
    </div>
    <button
      data-testid="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`pass FormControlLabelProps and FormHelperTextProps (checked false) 1`] = `
<DocumentFragment>
  <form
    action="#"
    data-testid="form"
  >
    <div
      class="MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          label="Checkbox"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-12"
              data-indeterminate="false"
              data-testid="checkbox"
              name="checked"
              type="checkbox"
              value="false"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Terms and Conditions
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      >
        Read the terms and conditions carefully.
      </p>
    </div>
    <button
      data-testid="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`renders as formik component 1`] = `
<DocumentFragment>
  <form
    action="#"
    data-testid="form"
  >
    <div
      class="MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-4"
              data-indeterminate="false"
              name="checked"
              type="checkbox"
              value="false"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        />
      </label>
      <p
        class="MuiFormHelperText-root"
      />
    </div>
    <button
      data-testid="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;
